core/arch/riscv/riscv.mk

```diff
@@ -35,6 +35,18 @@ endif
 
 CFG_MAX_CACHE_LINE_SHIFT ?= 6
 
+ifeq ($(CFG_RISCV_S_MODE),y)
+$(info Info: Builing core for S-Mode)
+CFG_RISCV_M_MODE := n
```

Comment:
$(call force,CFG_RISCV_M_MODE,n) and same below for `CFG_RISCV_S_MODE`.

That's in commit "riscv: core: riscv.mk: select privilege mode of OP-TEE core". With that fixed, the commit is:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Reply:
This will allow selecting the mode from command line ? `make ... CFG_RISCV_S_MODE=y`
The default is building for M-Mode, but Spike supports (M/U or M/S/U), or should modify the config in case we want to build for S-Mode ?

Reply:
> This will allow selecting the mode from command line ?

Yes. As I understand it, `CFG_RISCV_S_MODE` and `CFG_RISCV_M_MODE` are mutually exclusive. So if one is enabled (`=y`), then the other must be disabled (either undefined, empty or `=n`). By using `$(call force,...)` we make sure that such an inconsistent configuration is detected. For example `make CFG_RISCV_S_MODE=y CFG_RISCV_M_MODE=y` will report an error. With your proposed code (`:=`), it would report nothing and `CFG_RISCV_M_MODE` would still be `y` due to the higher priority of the command line over `:=`.

There is a very simple rule for `CFG_` in OP-TEE Makefiles: you set them either with `?=` or with `$(call force...)`. Any other way is a mistake. (That's only for `CFG_` ,other variables may be a bit different).

See: https://optee.readthedocs.io/en/latest/building/gits/optee_os.html#configuration-variables (we should perhaps add the above rule more clearly there...)

Reply:
Yes understood, the issue comes here:

```
$ make ARCH=riscv PLATFORM=spike CROSS_COMPILE64=riscv64-linux-gnu-  CFG_RISCV_S_MODE=y
Info: Builing core for S-Mode
core/arch/riscv/riscv.mk:40: *** CFG_RISCV_M_MODE is set to 'y' (from file) but its value must be 'n'.  Stop.
```
In platform conf.mk:
```
CFG_RISCV_M_MODE ?= y
```
Command line supposed to have more priority than file ?

Reply:
Sorry it is indeed a bit more complicated because the two variables are related. Please try the following:
```
CFG_RISCV_M_MODE ?= y
ifeq ($(CFG_RISCV_M_MODE),y)
ifeq ($(CFG_RISCV_S_MODE),y)
$(error CFG_RISCV_M_MODE and CFG_RISCV_S_MODE cannot be both 'y')
else
$(call force,CFG_RISCV_S_MODE,n)
endif
endif
ifeq ($(CFG_RISCV_S_MODE),y)
$(call force,CFG_RISCV_M_MODE,n)
endif
ifneq (y,$(call cfg-one-enabled,CFG_RISCV_M_MODE M CFG_RISCV_S_MODE))
$(error Either CFG_RISCV_M_MODE or CFG_RISCV_S_MODE must be 'y')
endif
```

With this, the command to build M mode is `make` and for S mode `make CFG_RISCV_S_MODE=y CFG_RISCV_M_MODE=n`. And hopefully invalid combinations should be caught too.